Add coap support

drasko opened the pull request:
Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

drasko committed:
Add coap support

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on cmd/coap/main.go:
```diff
@@ -0,0 +1,96 @@
+	NatsURL string
+}
+
+var logger *zap.Logger = nil
```
Avoid this, it is an unnecessary initialization. We have this initialization on 2 more places in normalizer service as well, and it should be fixed completely.

drasko replied:
How should we pass logger instance to the other modules? Through function parameter?

mijicd commented on cmd/coap/main.go:
```diff
@@ -0,0 +1,96 @@
+
+	nats.StoreConnection(nc)
+
+	// Initialize map of Observers
```
Drop this comment, the line is self-explanatory.

mijicd commented on cmd/coap/main.go:
```diff
@@ -0,0 +1,96 @@
+	// Initialize map of Observers
+	adapter.ObsMap = make(map[string][]adapter.Observer)
+
+	// Subscribe to NATS
```
Drop this one as well.

mijicd commented on cmd/coap/main.go:
```diff
@@ -0,0 +1,96 @@
+	errs := make(chan error, 2)
+
+	go func() {
+		// Serve CoAP
```
This one as well.

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+	broker "github.com/nats-io/go-nats"
+)
+
+type (
```
No need for multi-type definition. "Inline" it:
```
type Observer struct { ... }
```

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+	}
+)
+
+// Map of observers
```
This is not a valid godoc comment. Moreover, should this variable be exported?

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+// Map of observers
+var ObsMap map[string][]Observer
+
+// sendMessage function
```
This is not exported function, and its name is quite clear without repeating it in the comment.

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+		return res
+	}
+
+	// Channel ID
```
Unnecessary.

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+	// Channel ID
+	cid := mux.Var(m, "channel_id")
+
+	// Publish message via NATS
```
This can be inlined as well:

```
n := normalizer.Message{
    Channel: cid,
    ...
}
```

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+	m := normalizer.Message{}
+	if len(nm.Data) > 0 {
+		if err := json.Unmarshal(nm.Data, &m); err != nil {
+			println("Can not decode adapter msg")
```
Please avoid `println`.

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+
+	if err := nats.Send(n); err != nil {
+		if m.IsConfirmable() {
+			res.Payload = []byte("{\"error\": \"cannot publish\"}")
```
What is the purpose of `res.Payload`?

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+
+}
+
+// observeMessage function
```
See the remark for sendMessage.

mijicd commented on coap/message.go:
```diff
@@ -0,0 +1,184 @@
+		Message: m,
+	}
+
+	if m.Option(coap.Observe) != nil {
```
Refactor & simplify the following `if-else`.

mijicd commented on coap/server.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+// NotFound Handler - useful for ACK-EMPTY
+func notFoundHandler(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
```
See the remark for sendMessage.

mijicd commented on coap/adapter.go:
```diff
@@ -0,0 +1,57 @@
+	m := writer.RawMessage{}
+	if len(nm.Data) > 0 {
+		if err := json.Unmarshal(nm.Data, &m); err != nil {
+			println("Can not decode adapter msg")
```
No println's.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+		message: m,
+	}
+
+	if m.Option(coap.Observe) != nil {
```
Consider splitting into multiple methods to simplify this piece of code.

drasko committed:
Simplify functions

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+	cid := mux.Var(m, "channel_id")
+
+	// Publish message via NATS
+	n := writer.RawMessage{}
```
Inline this, and extract protocol as constant.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+
+	if err := nats.Send(n); err != nil {
+		if m.IsConfirmable() {
+			res.Payload = []byte("{\"error\": \"cannot publish\"}")
```
What is payload? Is this something you return to the clients? If so, drop it, since no other protocols perform the same action, and stuff like `{"error":"cannot publish"}`, or `{"res":"sent"}` provide absolutely no more information to the clients than bare response status.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	if m.IsConfirmable() {
+		res.Payload = []byte("{\"res\": \"observing\"}")
```
See the remark above.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	if m.Option(coap.Observe) != nil {
+		if value, ok := m.Option(coap.Observe).(uint32); ok {
```
No need for nested if. It can be avoided by moving `value` check to the outer one.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,154 @@
+// observeMessage adds client to the observers map
+func (ca *CoAPAdapter) observeMessage(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
+	log.Printf("Got message in observeMessage: path=%q: %#v from %v", m.Path(), m, a)
+	var res *coap.Message = nil
```
No need for `nil` initialization.

mijicd commented on coap/nats/messages.go:
```diff
@@ -0,0 +1,27 @@
+const topic string = "msg.coap"
+
+// Stored NATS connection
+var snc *broker.Conn
```
This way of initializing nats publisher is inconsistent with the rest of the platform.

drasko committed:
Initialize properly NATS in CoAP adapter

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on coap/adapter.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// MsgHandler functions is a handler for messages recieved via NATS
+func (ca *CoAPAdapter) MsgHandler(nm *broker.Msg) {
```
Either use it as anonymous, or name it properly.

drasko replied:
It is used twice: https://github.com/mainflux/mainflux/blob/coap/cmd/coap/main.go#L39-L40, I would like to avoid inlining two times.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,155 @@
+
+	if len(m.Payload) == 0 {
+		if m.IsConfirmable() {
+			res.Payload = []byte("{\"res\": \"Error: msg len can not be 0\"}")
```
See the remarks regarding previous payload occurences.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,155 @@
+	return res
+}
+
+// registerObserver functions adds observer struct to the observers map
```
Private methods does not require godoc. On top of this, function is named properly, hence, this doc only clutters the code.

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	if value, ok := m.Option(coap.Observe).(uint32); ok {
+		if value == 0 {
```
Nested if can be removed by moving `value` check to the outer if clause.

```
if value, ok := m.Option(coap.Observe).(uint32); ok && value == 0 {
    ca.registerObserver(o, cid)
} else {
    ca.deregisterObserver(o, cid)
}
```

mijicd commented on coap/messages.go:
```diff
@@ -0,0 +1,155 @@
+		msg := *(v.message)
+		msg.Payload = n.Payload
+
+		log.Printf("ca.obsMap[cid] = %v", v)
```
As general remark, check whether or not zap supports log level, since most of the logs you're making should have logging level `DEBUG` or `TRACE`.

drasko replied:
Zap has it, but this is highly experimental and WIP mode, I prefer that still some info is printed out for debug. This will be changed later.

mijicd committed:
Ensure codestyle adherence

Simplified code where possible. Fixed golint suggestions regarding the
missing godoc comments and unnecessary initialized variables.

Signed-off-by: Dejan Mijic <dejan@mainflux.com>

drasko committed:
Remove comments, clean code

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>

mijicd commented on cmd/coap/main.go:
```diff
@@ -0,0 +1,84 @@
+func connectToNats(cfg *config, logger *zap.Logger) *broker.Conn {
+	nc, err := broker.Connect(cfg.NatsURL)
+	if err != nil {
+		println(logger)
```
Remove println statements.

drasko committed:
Remove println

Signed-off-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
